ejb3/src/main/java/org/jboss/as/ejb3/timerservice/TimerImpl.java

```diff
@@ -555,31 +530,22 @@ public void unlock() {
         inUseLock.release();
     }
 
-    /**
-     * A {@link javax.ejb.Timer} is equal to another {@link javax.ejb.Timer} if their
-     * {@link javax.ejb.TimerHandle}s are equal
-     */
     @Override
-    public boolean equals(Object obj) {
-        if (obj == null) {
-            return false;
-        }
-        if (this.handle == null) {
-            return false;
-        }
-        if (obj instanceof TimerImpl == false) {
-            return false;
-        }
-        TimerImpl otherTimer = (TimerImpl) obj;
-        return this.handle.equals(otherTimer.getTimerHandle());
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (this.getClass() != o.getClass()) return false;
```

Comment:
The class is not final and you're switching from instanceof to getClass. This might cause trouble for derived classes.

Reply:
this is to make sure `this` and `other` are of the exactly the same type. We don't want to compare an instance of super type against a derived type. With `instanceof`, if `this` is `TimerImpl` and `other` is `SubTimerImpl`, `other instanceof TimerImpl` would eval to true and continue, which is not what we want.

Reply:
thanks for the explanation